src/make/ppx_make.ml

```
@@ -0,0 +1,247 @@
+      (* [@default _ ] -> Optional *)
+      | Some _, _ -> Optional name.txt, ty
+      (* `option` type -> Optional *)
+      | _, [%type: [%t? a'] option] -> Optional name.txt, a'
+      (* `list` type   -> Optional *)
+      | _, [%type: [%t? _] list] -> Optional name.txt, ty
+      (* Regular field -> Labelled *)
+      | _ -> Labelled name.txt, ty
+     ;;
+     
+     let create_make_sig ~loc ~ty_name ~tps label_decls =
+       let record = Construct.apply_type ~loc ~ty_name ~tps in
+       match Annotations.find_main label_decls with 
+        | Error e -> List.map(fun e ->  psig_extension ~loc (e) [] ) e
+        | Ok (main_arg, label_decls) -> 
+       let types = List.map label_arg label_decls in
+       let add_unit types = (
+         Nolabel, 
+         Ast_helper.Typ.constr ~loc { txt = Lident "unit"; loc } []
+        )::types in
+       let types = match main_arg with 
+        | Some { pld_type ; _ } 
+            -> (Nolabel, pld_type)::types
+        | None when Check.is_optional types -> add_unit types
+        | None -> types
+       in
+       let t = Construct.lambda_sig ~loc types record in
+       let fun_name = "make_" ^ ty_name in
+       [Construct.sig_item ~loc fun_name t]
+     ;;
+
+     let derive_per_td (td : type_declaration) : signature =
+       let { ptype_name = { txt = ty_name; loc }
+           ; ptype_private = private_
+           ; ptype_params
+           ; ptype_kind
+           ; _
+           }
+         =
+         td
+       in
+       let tps = List.map (fun (tp, _variance) -> tp) ptype_params in
+       match ptype_kind with
+       | Ptype_record label_decls ->
+        if private_ = Public then  
+         let derived_item = create_make_sig ~loc ~ty_name ~tps label_decls in
+        derived_item
+      else 
+        [psig_extension ~loc 
+          (Location.error_extensionf ~loc "We cannot expose functions that explicitly create private records.") [] ]
+       | _ -> []
```

If I understand correctly, this case is never supposed to be hit. Is that right? If so, I think instead of silently not doing anything, it would be better to return an error (by embedding the error as an extension node like always). It could be something along the lines "There has been an internal error while encountering a non-record type".

---

Checking for non-record types happens earlier in the is_derivable function, and there's a corresponding error: "Unsupported use of make (you can only use it on records)" @pitag-ha

---

^ I'm tagging you to ask if we should still add that error msg if there is a similar one we explicitly check for earlier

---

It's up to you. I'd still add that error message for the following reason: if at some point in the future the code is changed and, in particular, the part of the code checking for non-record types earlier is removed (i.e. a small "bug" is introduced), then it would be good to have an error message that makes clear that there was an unexpected error in the deriver. With the current implementation, the deriver will simply ignore that type. Given that the deriver isn't supposed to be used on non-records in the first place though, it's also ok to leave it as is, if you prefer that.

---

Ok just tried to change it and understood the need for `[]` in the tests with recursive types where there is a type that is not a record type: as long as at least 1 of the recursive types is a record, the deriver should still work on that record type. 

This point in the code is the case looking at the other non-record types in the recursive definition, and we don't actually want to throw an error.

---

Left it as is but maybe a clarifying comment might be useful?

---

> and understood the need for [] in the tests with recursive types where there is a type that is not a record type: as long as at least 1 of the recursive types is a record, the deriver should still work on that record type.

Oooh, that's a very good point! Nice, your super complete test base has already proven very useful!

> Left it as is but maybe a clarifying comment might be useful?

Yes, I think that might be nice. Thanks!